### ISSUE TITLE
KC 21.1.1: Error shown intermittently while opening admin console page 

### ISSUE THREAD

**Issue by souravs17031999** [keycloak/keycloak#20196]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

admin/ui

### Describe the bug

Hi, 

When I open the admin console UI of keycloak which is currently deployed in k8s cluster, then intermittently it shows error (some JS script error) although it redirects to correct login page after few seconds (~1-3 seconds usually). 

![keycloak admin ui error](https://user-images.githubusercontent.com/33771969/236864222-4ed9716e-428c-4df6-8cd2-4581a2e712e0.PNG)

### Version

21.1.1

### Expected behavior

Expected behaviour is while opening admin console at ```/auth/admin/master/console/```, it should take to login page normally.

### Actual behavior

Actual behaviour is while opening admin console at ```/auth/admin/master/console/```, it shows page with error in JS script and then after few seconds takes to normal login page.

### How to Reproduce?

1. Deploy keycloak in a pod in cloud (k8s)
2. Simply open the admin console page at ```/auth/admin/master/console/```

### Anything else?

Not sure whether it's related to any network issues or anything but can be bad user experience.

**Comment by edewit** [comment]
I think this is triggered by a token that is no longer valid, when that error occurs we redirect to the login page.

**Comment by mungi** [comment]
Upload the HAR file that I reproduced in my test environment. Hopefully this will help you troubleshoot.
[keycloak.har.zip](https://github.com/keycloak/keycloak/files/11430995/keycloak.har.zip)

**Comment by souravs17031999** [comment]
Here's one more in case it helps (from my environment)
[iam-cid21.zip](https://github.com/keycloak/keycloak/files/11431910/iam-cid21.zip)

Btw, the request that it shows failed, from the network tab it shows 200 Ok.

**Comment by jonkoops** [comment]
Thanks for providing some detailed network logs, I think this confirms @edewit's suspicion that this indeed triggered by using a token that is somehow no longer valid. There are a couple of enhancements that I am seeing here that we could be doing to make this a better experience.

First off, when we [initialize the Admin Console](https://github.com/keycloak/keycloak/blob/025778fe9c745316f80b53fe3052aeb314e868ef/js/apps/admin-ui/src/main.tsx#L15) we are using the `check-sso` option, which does not force the user to be authenticated. This means that we only force the user to log in later in the bootstrapping process, such as retrieving the user information (which causes this error) or entering a route.

We can move this process into the bootstrapping phase so that the user is re-directed to the login screen pretty much immediately.

Secondly, the page we are seeing here is the error page for React Router. For this we should implement a properly styled error page (see #19307).

**Comment by jonkoops** [comment]
Another case where this seems to happen is when you log in as a user that does not have permission to make calls to the `whoami` endpoint. If you create a user that has no permissions and impersonate it this issue can be reproduced, but no redirect takes place.

**Comment by jonkoops** [comment]
@mungi @souravs17031999 were y'all authenticated with a user that does not have admin permissions before encountering this problem?

**Comment by mungi** [comment]
After scaling the Keycloak deployment from 0 to 3, I opened the admin page url directly from Chrome's Incognito window, meaning I was not logged in. There was no authentication process to reproduce this issue - it was a test environment with no special settings, just the admin default account.

**Comment by jonkoops** [comment]
Perhaps multiple instances of the Keycloak server are running and their tokens cannot be used interchangeably? I've merged #20251 so we should now always be checking the authentication status of the user before boostrapping the UI.

Can someone try out the [latest nightly](https://github.com/keycloak/keycloak/releases/tag/nightly) and verify this problem has been fixed or not?

**Comment by souravs17031999** [comment]
@jonkoops I was able to reproduce even with single instance of keycloak server (that's how First time I got it).
I will try to check the latest nightly.

**Comment by lrozenblyum** [comment]
@jonkoops I can confirm the problem is still happening occasionally with 22.0.3 release.

**Comment by jonkoops** [comment]
Hmmm, I am still unable to reproduce this locally. @keycloak/ui is anyone of you able to reproduce this?

**Comment by mannp** [comment]
> @jonkoops I can confirm the problem is still happening occasionally with 22.0.3 release. Single instance.

It is still happening for me too on a single instance of keycloak 22.0.3

Will try nightly.

**Comment by jonkoops** [comment]
Thanks, let me know if you manage to reproduce it there. If you have some steps to reproduce the issue that would be much appreciated.

**Comment by mannp** [comment]
> Thanks, let me know if you manage to reproduce it there. If you have some steps to reproduce the issue that would be much appreciated.

I'd love to help but updating to nightly has somehow reset my realms and I have lost them all :-/ (only master now exists).

**Comment by ssilvert** [comment]
@mannp If you download the nightly build and copy over the data directory from the old build then  you would be set.

Is your data lost such that you can no longer reproduce the problem?

**Comment by mannp** [comment]
> @mannp If you download the nightly build and copy over the data directory from the old build then  you would be set.
> 
> Is your data lost such that you can no longer reproduce the problem?

Hi @ssilvert I stupidly used nightly with my daily used config and roasted it.

Restored KC and my db from backups and I still don't have my clients any more.

I am not able to help you unfortunately, and not had chance to see how to get things back yet.

 Will certainly try and help once I get my mini drama sorted :-)

**Comment by ssilvert** [comment]
@mannp @lrozenblyum @souravs17031999 @mungi @antikalk 

Do any of you have a way to reproduce this on a recent build?  What is the status of this issue?

**Comment by mannp** [comment]
> @mannp @lrozenblyum @souravs17031999 @mungi @antikalk
> 
> Do any of you have a way to reproduce this on a recent build? What is the status of this issue?

Restore of db and config failed to restore my personal instance, seems clients have to be backed up separately (which for me was older) so I changed to another product.

My business instance has the issues, but it still works, so I will continue with it until I get some bandwidth to look into it further.

I am using a postgres db with both instances.

**Comment by ssilvert** [comment]
@mannp @lrozenblyum @souravs17031999 @mungi @antikalk

I just stumbled onto a way to reproduce this error.  I had an old bookmarked URL that appended `/#/realms/master`.  Full URL was `http://localhost:8180/admin/master/console/#/realms/master`

We no longer include the `realms` part of that URL.  So Keycloak was trying to find a realm named `realms`.  The result was a NullPointerException on the server side and the aforementioned error message in the browser.

Is this what you guys are seeing?  If so, the simple solution is to fix the (bookmarked?) URLs.  

But we should also handle this better in the admin console and give a better error message and a way to recover to the main page of the admin console app.

**Comment by mannp** [comment]
> @mannp @lrozenblyum @souravs17031999 @mungi @antikalk
> 
> I just stumbled onto a way to reproduce this error. I had an old bookmarked URL that appended `/#/realms/master`. Full URL was `http://localhost:8180/admin/master/console/#/realms/master`
> 
> We no longer include the `realms` part of that URL. So Keycloak was trying to find a realm named `realms`. The result was a NullPointerException on the server side and the aforementioned error message in the browser.
> 
> The correct URL should be something like, `http://localhost:8180/admin/master/console/#/master`
> 
> Is this what you guys are seeing? If so, the simple solution is to fix the (bookmarked?) URLs.
> 
> But we should also handle this better in the admin console and give a better error message and a way to recover to the main page of the admin console app.

Not for me @ssilvert, I go to the root fqdn for keycloak and it redirects me to login the console from there. 

I have no bookmarked urls at all for KC.

The only instance this might be the case is if an oidc client has realms in the url, I would have to check.

Are these potentially being cached and causing issues that way perhaps?

**Comment by ssilvert** [comment]
> I have no bookmarked urls at all for KC.
> 
> The only instance this might be the case is if an oidc client has realms in the url, I would have to check.
> 
> Are these potentially being cached and causing issues that way perhaps?

I'm not sure, but it could be something like that.  Since we don't have a way to reproduce, I'm thinking of closing this issue soon.

**Comment by ssilvert** [comment]
Closing.  We can reopen if someone can find a way to reliably reproduce the problem.

**Comment by lrozenblyum-iqmessenger** [comment]
@ssilvert @jonkoops 

I was able to reproduce it in such a way: load the page `/auth/admin/<realm name>/console/`
and press F5 multiple times in Google chrome (on Keycloak 22.0.5).

Not sure if it's the best way to reproduce but anyway at least something.

1. 
```
 React Router caught the following error during render TypeError: Failed to fetch
    at fetchWithError (/auth/resources/6hg4h/admin/keycloak.v2/assets/index-5be21fbf.js:67:121308)
    at Agent.requestWithParams (/auth/resources/6hg4h/admin/keycloak.v2/assets/index-5be21fbf.js:67:124017) {componentStack: '\n    at WhoAmIContextProvider (https://192.168.17.…min/keycloak.v2/assets/index-5be21fbf.js:58:8667)'}
```

2.
```
 TypeError: Failed to fetch
    at fetchWithError (/auth/resources/6hg4h/admin/keycloak.v2/assets/index-5be21fbf.js:67:121308)
    at Agent.requestWithParams (/auth/resources/6hg4h/admin/keycloak.v2/assets/index-5be21fbf.js:67:124017)
```

3. Same ex. as 2.

**Comment by jonkoops** [comment]
@lrozenblyum that version of Keycloak is no longer supported. If you can reproduce this on the latest version feel free to submit a new issue with reproducible steps.

**Comment by a1994sc** [comment]
@jonkoops I'm seeing this on `25.0.2` fresh install on k8s.

**Comment by ssilvert** [comment]
The cloud native team might want to look at this.

**Comment by shawkins** [comment]
> After that is up, scale keycloak statefulset to 3
> Try to login and get the Something went wrong  error

I tried a couple of times to reproduce this with the operator, but could not.

While the cluster members are in a not ready state, I see a 503 response in the browser. Once one of the instances is ready, I see the admin page without an error.

@a1994sc are you using the operator for the statefulsets / services / ingress?
